### PR TITLE
Modernizes SQL admin loading

### PR DESCRIPTION
**Pull request by Jordie0608**
Updates sql based admin loading to support excluding flags from ranks and to allow for tempminning. Since changes are largely piggybacking off the existing systems for loading admins, legacy loading is unchanged and still supported. Functionality for adding and removing individual verbs from ranks is removed. Any database errors when loading admins forces a revert to legacy loading.

When sql admins is enabled actions on the permission panel are prompted to be either written to the DB or just for the current round, where applicable. Additionally when sql is admins enabled the list of admins and ranks is reloaded from DB and recorded to `admins.txt` and `admin_ranks.txt` as a backup. This is also where updating `lastadminrank` is handled so only permanent changes are kept.

The interface for changing rank flags now uses @vuonojenmustaturska bitfield switcher, which has been modified to allow changing window height, width and slider color. The exclusion flag picker uses red sliders for clarity. For reference, the bitfield switcher looks like: https://github.com/tgstation/tgstation/pull/34078#issuecomment-359104142

`rights2text()` updated for current admin flags and to output both included and excluded flags.
`code\modules\admin\permissionverbs\permissionedit.dm` moved to `code\modules\admin\permissionedit.dm` as it was the only file there.

All three admin tables are modified to remove unused columns and improve the remaining. `admin_log` now records operation type per log for easier searching, however note there's currently no interface for reading the log.

A python script is included to handle importing admins and ranks from `admins.txt` and `admin_ranks.txt`

General functionality tested to work, pending more testing to root out possible edge cases.
@MrStonedOne 
@nfreader

### CODE REVIEW

**Comment by nfreader** [comment]
`happening`

**Comment by Shadowlight213** [comment]
Add a new permission flag for actually saving changes to the db for the headmins, or major :-1: on this.
Actually, :-1: on letting the admins list be permanently modified from within the game period.
We already had a headmin byond account be compromised once.


The old panel provided trainers the ability to tempmin people, while restricting permanent changes to the txt to people with remote. Any changes would get reset the next round.
Now, a rogue admin could remove everyone from the txt, and it would last until someone with remote managed to get on to manually reset to legacy, then fix the txt.
Or, the more likely scenario where a trainer accidentally makes a candidacy permanent without noticing. This causes more issues than it solves.

**Comment by kevinz000** [comment]
add a database write flag and make it an option to make changes permanent.
also while you're at it can we get a flag for tgdb access?

**Comment by Shadowlight213** [comment]
Direct tgdb access, as in bypass the proc call wrapper?
That's an awful idea.

**Comment by nfreader** [comment]
tgdb access is handled by forum groups/byond linking at the moment (unless I am horribly mistaken). I don't think this update will affect that though, this would only handle in-game stuff.

**Comment by MrStonedOne** [comment]
Jordie, could you potentially make db adding permissions be flag based, such that one can restrict what flags they are allowed to add and/or restrict what roles they are allowed to add somebody to?

**Comment by Jordie0608** [comment]
@Shadowlight213 New permission flag sure. However an account being compromised is a human issue; Any form of access control is susceptible to identity theft and impersonation, each scenario you describe could happen today with anybody that has remote and it can't feasibly be protected against. The solution to that is better data security.

@MrStonedOne Yes, but not sure how to best implement it; my first idea is irc-like rank hierarchy letting users edit up to the level below them, but that doesn't work too well with hoe generously our permission flags are distributed. Thoughts?

**Comment by MrStonedOne** [comment]
No need for hierarchy, just a second set of flags for the rank stating what exact flags they are allowed to add to other people, than only allow them to add ranks that only have those flags.

**Comment by MrStonedOne** [comment]
Also, store a cache of admins in data as json, leave admins.txt alone, continue to read from it. than we can keep headmin ranks in the .txt that can't be removed by the database as a failsafe for admins getting hacked

**Comment by Jordie0608** [comment]
- `R_DBRANKS` added, only temporary permission panel changes can be made without it.
- `can_edit_rights` added to admin ranks, rights not in this flag are blocked from being changed in the permissions panel. Blocked rights are greyed out with a padlock padlock
- Bitfield switcher edited to allow for locking flags.
- `can_edit_rights` is assigned in `admin_ranks.txt` by `*` prefix.
- Admins and ranks are now loaded from txt first and will override any loaded from the database.
- `PROTECT_LEGACY_ADMINS` and `PROTECT_LEGACY_RANKS` configs added that prevent editing the rank and permissions respectively of an admin or rank that was loaded from txt.
- Backup is now a json in `data/`
- `admin_ranks` `log` column size doubled to accommodate extra logging from `can_edit_rights` (maybe it should just not parse the flag to text?)

More testing to follow when I get up.

**Comment by Jordie0608** [comment]
Ok all good

**Comment by optimumtact** [comment]
does this need coordination on merge?

**Comment by Jordie0608** [comment]
It doesn't need it, but it'll be safer so mso can confirm the db import works.

File: SQL/database_changelog.txt [hunk 1]
```diff
@@ -1,13 +1,47 @@
...
+INSERT INTO `schema_revision` (`major`, `minor`) VALUES (4, 1);
 or
-INSERT INTO `SS13_schema_revision` (`major`, `minor`) VALUES (4, 0);
+INSERT INTO `SS13_schema_revision` (`major`, `minor`) VALUES (4, 1);
 
 In any query remember to add a prefix to the table names if you use one.
 
+----------------------------------------------------
+
+3 February 2018, by Jordie0608
```
Comment (MrStonedOne) [inline SQL/database_changelog.txt]: I'd like to see us start to add version numbers in this header, as otherwise if you are behind a few updates you can't know what updates to apply.

Comment (Jordie0608) [reply SQL/database_changelog.txt]: I'll just do it for all the headers since versioning.

File: SQL/database_changelog.txt [hunk 2]
```diff
@@ -1,13 +1,47 @@
...
+	CHANGE COLUMN `datetime` `datetime` DATETIME NOT NULL AFTER `id`,
+	CHANGE COLUMN `adminckey` `adminckey` VARCHAR(32) NOT NULL AFTER `datetime`,
+	CHANGE COLUMN `adminip` `adminip` INT(10) UNSIGNED NOT NULL AFTER `adminckey`,
+	ADD COLUMN `operation` ENUM('add admin','remove admin','change admin rank','add rank','remove rank','change rank flags') NOT NULL AFTER `adminip`,
+	CHANGE COLUMN `log` `log` VARCHAR(1000) NOT NULL AFTER `operation`;
+
+ALTER TABLE `admin_ranks`
+	CHANGE COLUMN `rank` `rank` VARCHAR(32) NOT NULL FIRST,
+	CHANGE COLUMN `flags` `flags` SMALLINT UNSIGNED NOT NULL AFTER `rank`,
+	ADD COLUMN `exclude_flags` SMALLINT UNSIGNED NOT NULL AFTER `flags,`
```
Comment (MrStonedOne) [inline SQL/database_changelog.txt]:
```
`flags,`
```

File: SQL/database_changelog.txt [hunk 3]
```diff
@@ -1,13 +1,47 @@
...
+	ADD COLUMN `can_edit_flags` SMALLINT(5) UNSIGNED NOT NULL AFTER `exclude_flags`
```
Comment (MrStonedOne) [inline SQL/database_changelog.txt]: Missing comma

File: SQL/admin_import_2018-02-03.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+with open("..\\config\\admin_ranks.txt") as rank_file:
+    previous = 0
+    for line in rank_file:
+        if line.strip():
+            if line.startswith("#"):
+                continue
+            matches = re.match("(.+)\\b\\s+=\\s*(.*)", line)
+            flags = parse_text_flags(matches.group(2), previous)
+            previous = flags[0]
+            cursor.execute("INSERT INTO {0} (rank, flags, exclude_flags, can_edit_flags) VALUES ('{1}', {2}, {3}, 0)".format(ranks_table, matches.group(1), flags[0], flags[1]))
```
Comment (MrStonedOne) [inline SQL/admin_import_2018-02-03.py]: image

This is with the default repo admins and ranks.

I don't think this is correct.

**Comment by MrStonedOne** [comment]
image

1: Remove `!localhost!`'s `+DBRANKS` i
2: Make `!localhost!` not apply if the connecting ckey already has admin access since now its possible for `!localhost!` to have less permissions. (as seen here)
3: We should likely make a config to enable the localhost rank that defaults to enable in the .txt as i could see it being a compounding security risk if somebody figures out an exploit to relay tcp packets in windows or byond and uses it to spoof a localhost connection. In a production environment ip auth should be a secondary security measure, not a primary form of authentication. (looking at you memcached)

**Comment by MrStonedOne** [comment]
```
## Comment this out if you want to use the SQL based admin system, the legacy system uses admins.txt.
## You need to set up your database to use the SQL based system.
## This flag is automatically enabled if SQL_ENABLED isn't
#ADMIN_LEGACY_SYSTEM

#Uncomment this to stop any admins loaded by the legacy system from having their rank edited by the permissions panel
PROTECT_LEGACY_ADMINS

#Uncomment this to stop any ranks loaded by the legacy system from having their flags edited by the permissions panel
PROTECT_LEGACY_RANKS
```

```
Error: You do not have sufficient rights to do that. You require one of the following flags: +PERMISSIONS.


Current Admins:
        MrStonedOne is a Admin Candidate - Lobby
```


I don't think the protection functionality is working, unless its intended for legacy people to be able to edit their rank or delete themselves.

File: code/modules/admin/admin_ranks.dm [hunk 5]
```diff
@@ -100,165 +102,170 @@ GLOBAL_PROTECT(admin_ranks)
...
 	#endif
 
-
 /proc/load_admins()
+	var/dbfail
+	if(CONFIG_GET(flag/admin_legacy_system) && !SSdbcore.Connect())
+		message_admins("Failed to connect to database while loading admins. Loading from backup.")
+		log_sql("Failed to connect to database while loading admins. Loading from backup.")
+		CONFIG_SET(flag/admin_legacy_system, TRUE)
+		dbfail = 1
```
Comment (MrStonedOne) [inline code/modules/admin/admin_ranks.dm]: I can see this being an annoyance if the issue is later fixed and there is no way to attempt to revert back to dbmins until the next round.

File: code/modules/admin/admin_ranks.dm [hunk 6]
```diff
@@ -100,165 +102,170 @@ GLOBAL_PROTECT(admin_ranks)
...
+		var/rights = rights2text(R.rights,"\n\t\t")
 		if(rights)
 			msg += "\t\t[rights]\n"
 	testing(msg)
 	#endif
 
-
 /proc/load_admins()
+	var/dbfail
+	if(CONFIG_GET(flag/admin_legacy_system) && !SSdbcore.Connect())
```
Comment (MrStonedOne) [inline code/modules/admin/admin_ranks.dm]: I think you are missing a `!`

File: code/modules/admin/permissionedit.dm [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+	var/use_db
+	var/task = href_list["editrights"]
+	var/skip
+	if(task == "activate" || task == "deactivate")
+		skip = 1
+	if(CONFIG_GET(flag/admin_legacy_system) && CONFIG_GET(flag/protect_legacy_admins) && task == "rank")
+		if(admin_ckey in GLOB.protected_admins)
+			to_chat(usr, "<span class='admin prefix'>Editing the rank of this admin is blocked by server configuration.</span>")
+			return
+	if(CONFIG_GET(flag/admin_legacy_system) && CONFIG_GET(flag/protect_legacy_ranks) && task == "permissions")
```
Comment (MrStonedOne) [inline code/modules/admin/permissionedit.dm]: `!`

File: code/modules/admin/permissionedit.dm [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+	var/datum/asset/permissions_assets = get_asset_datum(/datum/asset/simple/permissions)
+	permissions_assets.send(src)
+	var/admin_ckey = ckey(href_list["ckey"])
+	var/datum/admins/D = GLOB.admin_datums[admin_ckey]
+	var/use_db
+	var/task = href_list["editrights"]
+	var/skip
+	if(task == "activate" || task == "deactivate")
+		skip = 1
+	if(CONFIG_GET(flag/admin_legacy_system) && CONFIG_GET(flag/protect_legacy_admins) && task == "rank")
```
Comment (MrStonedOne) [inline code/modules/admin/permissionedit.dm]: `!`

**Comment by Jordie0608** [comment]
I could go further and make it so !localhost! can be defined in `admin_ranks.txt` and use the permissions listed for it there, or if not present `+EVERYTHING -DBRANKS *EVERYTHING`
